Yannick491 - xgboost factor levels

yvanhuele opened the pull request:
Changes:
- Created a modelInfo object in supervised model development
   - fitLogit and factorLevels are now attributes of modelInfo
   - all models (including xgboost) save a modelInfo object
   - moved the model saving to the supervised model development parent class and added a modelName attribute to each model to allow model saving to occur in the parent class while still having different file names for different models
- Perform factor level coercion for xgboost
- Added unit tests related to factor level coercion for xgboost
- Added a function called ignoreSpecWarn to common testing utils which allows us to suppress specific warnings we know will trigger in the unit tests, but are unrelated to the functionality being tested.

mmastand commented on tests/testthat/test-xgboost-deploy.R:
```diff
@@ -73,3 +79,226 @@ test_that("Top categories are correctly parsed", {
+  xDeployChar <- capture.output(boostDChar$deploy())
+  xDfChar <- capture.output(outDfChar <- boostDChar$getOutDf())
+  
+  # Check that we get the same results when using factors/characters
```
Excellent test.

mmastand replied:
Actually, all of them are great. If these pass, I'm happy with xgb. Until we find more bugs of course 😄

mmastand commented on tests/testthat/test-xgboost-deploy.R:
```diff
@@ -73,3 +79,226 @@ test_that("Top categories are correctly parsed", {
+# factor level tests.  In particular, a toy multiclass dataset is generated
+# with the option of treating strings as factors or characters (determined by
+# the asFactor parameter).
+getToyXGBDevelopParams = function(asFactor) {
```
Good use of a function to make these tests independent.

yvanhuele commented on R/lasso-development.R:
```diff
@@ -176,7 +176,7 @@ LassoDevelopment <- R6Class("LassoDevelopment",
     # Fit model and lamda values
     fitGrLasso = NA,
-    fitLogit = NA,
+    modelName = "lasso",
```
It will require a little bit of work because the model names also have to be specified in deployment, I'll see what I can do.  Looking ahead, I think there's more functionality of this type we might want to add (see Issue #514)

yvanhuele commented on R/supervised-model-deployment.R:
```diff
@@ -244,13 +246,42 @@ SupervisedModelDeployment <- R6Class("SupervisedModelDeployment",
+      load(modelInfoFile)  # Get model info
+      self$modelInfo <- modelInfo
+      load(fitObjFile) # Produces fit object (for probability)
+      self$fitObj <- fitObj
```
I wanted to pass the function loadModelAndInfo an additional parameter to directly assign fitObj to fitRF, fitGrLasso, etc. but couldn't figure out how to pass a reference (rather than a copy) to a function (see below) so I added a fitObj variable to SupervisedModelDeployment and assign this to fitRF, fitGRLasso, etc. in each of the individual classes.

Example code showing how R passes function arguments as copies rather than references:
```
unknown_fruit <- NA

assign_banana = function(fruit) {
  fruit <- "Banana"
}

assign_banana(unknown_fruit) # doesn't change the value of unknown_fruit
is.na(unknown_fruit) # returns TRUE
```

yvanhuele commented on R/lasso-deployment.R:
```diff
@@ -537,26 +536,21 @@ LassoDeployment <- R6Class(
+      private$modMat <- self$fitObj$modMat
+      private$modFmla <- self$fitObj$modFmla
+      self$fitObj$modMat <- NULL
+      self$fitObj$modFmla <- NULL
```
I'm not what the purpose of lines 552 and 553 is, but I wanted to follow the old code as closely as possible

yvanhuele commented:
New changes:
- made modelName an optional parameter to supervised model development and deployment so users can save their models under a custom name.  This partially addresses #514.
- added unit test which tests saving a model with a custom name
- added more comments explaining the tests in test-xgboost-deploy

mmastand commented:
This is showing an error on the OSX travis build, but the travis site shows that it's working. Merging! Great work, Yannick.
